analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs

```diff
@@ -0,0 +1,124 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[ExportCodeFixProvider(LanguageNames.CSharp)]
+public sealed class CommentedOutCodeCodeFix : SonarCodeFix
+{
+    public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CommentedOutCode.DiagnosticId);
+
+    public override FixAllProvider GetFixAllProvider() => null;
+
+    protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
+    {
+        var diagnostic = context.Diagnostics.First();
+        var comment = root.FindTrivia(diagnostic.Location.SourceSpan.Start, findInsideTrivia: true);
+
+        if (comment.IsKind(SyntaxKind.SingleLineCommentTrivia) || IsCode(comment))
+        {
+            context.RegisterCodeFix(CodeAction.Create(
+                CommentedOutCode.MessageFormat,
+                c => new Context(comment).ChangeDocument(context.Document, root)),
+                diagnostic);
+        }
+        return Task.CompletedTask;
+    }
+
+    private static bool IsCode(SyntaxTrivia trivia) =>
+        trivia.GetLineNumbers().Count() == 1
+        || Array.TrueForAll(Lines(trivia), IsCode);
+
+    private static bool IsCode(string line)
+    {
+        var trimmed = line.Trim();
+        return trimmed == "/*"
+            || trimmed == "*/"
+            || string.IsNullOrEmpty(trimmed)
+            || CommentedOutCode.IsCode(line);
+    }
+
+    private static string[] Lines(SyntaxTrivia trivia) =>
+        trivia.ToFullString().Split(MetricsBase.LineTerminators, StringSplitOptions.None);
+
+    internal sealed class Context
+    {
+        public Context(SyntaxTrivia comment)
+        {
+            Comment = comment;
+            Leading = Token.LeadingTrivia.ToList();
+            Trailing = Token.TrailingTrivia.ToList();
+
+            IsTrailing = Trailing.Remove(comment);
+
+            if (!IsTrailing)
+            {
+                Leading.Remove(comment);
+            }
+        }
+
+        public bool IsTrailing { get; }
+        public SyntaxTrivia Comment { get; }
+        public List<SyntaxTrivia> Leading { get; }
+        public List<SyntaxTrivia> Trailing { get; }
+        public SyntaxToken Token => Comment.Token;
+
+        public Task<Document> ChangeDocument(Document document, SyntaxNode root) =>
+            Task.FromResult(document.WithSyntaxRoot(root.ReplaceToken(Token, NewToken)));
+
+        public SyntaxToken NewToken =>
+            IsTrailing ? NewTrailing() : NewLeading();
+
+        private SyntaxToken NewTrailing()
+        {
+            var trailing = ShareLine(Token, Comment)
+                ? Trailing.Where(t
```

Comment:
Align it like this, as per the [style guide](https://github.com/SonarSource/sonar-dotnet/blob/master/docs/coding-style.md#multi-line-statements):

```csharp
? Trailing.Where(t => !ShareLine(t, Comment)
                      || (t.IsKind(SyntaxKind.EndOfLineTrivia) && ShareLine(t, Comment))).ToList()

```

Reply:
Moved the Where constraint to a method. (I really don't like this style guide by the way.)

Reply:
I see you refactored it into a `XOR`. Semantically, this is not the same operation. 

If you make the [Map](https://en.wikipedia.org/wiki/Karnaugh_map) of the operations, you can see this:

c1 = `!ShareLine`
c2 = `IsEndOfLineTrivia`

|  c1 | c2  | c1 + c1!*c2  | c1 XOR c2  |  identical |
|---|---|---|---|---|
| 0  | 0  |  0  | 0  | ✅   |
| 0  | 1  |  1 |  1 | ✅   |
| 1  | 0  |  1 |  1 | ✅   |
| 1  | 1  |  1 |  0 | ❗   |

Basically, `before` you were keeping the trivia that "did not share a line **AND** were `EndOfLine`", but `after` the refactor, you do not.

Is this intended?

Reply:
Note: it is not my favorite styling guide either, but I like first-class methods-as-arguments in `Where` clauses like you did more anyways 😄

Reply:
It was an XOR: !a || (a && b) => !a ^ b

Reply:
I do not think these two expressions are the same.

Let's say you take the case of `(!a, b)` being `true`:

- `!a || (a && b)` => `true`
- `!a ^ b` => `false`

Am I missing something?

Reply:
No, you're not. It was late. I wasn't thinking straight. Anyway, it should be an OR, not an XOR. that being said, the fact that changing from OR to XOR did not make any test fail, suggests we miss a scenario.

Reply:
That's possible. Or it's something that can never happen, based on the way the AST is structured.
I think the simplified OR is pretty readable.